File: BlackFireIM/Example/Home/View/BFSparkLoadingView.m
//
//  BFSparkLoadingView.m
//  BlackFireIM
//
//  Created by benny wang on 2021/4/20.
//

#import "BFSparkLoadingView.h"
#import "BFHeader.h"

@interface BFSparkLoadingView()

@property(nonatomic,strong) UIImageView *bgImg;
@property(nonatomic,strong) UIImageView *topImg;
@property(nonatomic,strong) UIImageView *maskImg;
@property(nonatomic,strong) UIImageView *headImg;

@end
@implementation BFSparkLoadingView

- (instancetype)initWithFrame:(CGRect)frame
{
    if(self = [super initWithFrame:frame]) {
        self.bgImg = [UIImageView new];
        self.bgImg.image = [UIImage imageNamed:@"loadingbg"];
        [self addSubview:self.bgImg];
        
        self.topImg = [UIImageView new];
        self.topImg.image = [UIImage imageNamed:@"loadingtop"];
        [self addSubview:self.topImg];
        
        self.maskImg = [UIImageView new];
        self.maskImg.image = [UIImage imageNamed:@"loadingheadmask"];
        [self addSubview:self.maskImg];
        
        self.headImg = [UIImageView new];
        self.headImg.image = [UIImage imageNamed:@"loadinghead"];
        [self addSubview:self.headImg];
        self.alpha = 0;
    }
    return self;
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    self.bgImg.frame = CGRectMake(0, 0, Screen_Width, Screen_Width);
    self.bgImg.center = self.center;
    self.topImg.frame = CGRectMake(0, 0, Screen_Width, Screen_Width);
    self.topImg.center = self.center;
    self.maskImg.frame = CGRectMake(0, 0, 97, 97);
    self.maskImg.center = self.center;
    self.headImg.frame = CGRectMake(0, 0, 41, 55);
    self.headImg.center = self.center;
}

- (void)beginAnimating
{
    self.alpha = 0;
    [UIView animateWithDuration:0.25 animations:^{
        self.alpha = 1;
        
        } completion:^(BOOL finished) {
            CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"transform.rotation.z"];
            animation.fromValue = [NSNumber numberWithFloat:0.f];
            animation.toValue = [NSNumber numberWithFloat: M_PI *2];
            animation.duration = 3;
            animation.autoreverses = NO;
            animation.fillMode = kCAFillModeForwards;
            animation.repeatCount = MAXFLOAT;
            animation.removedOnCompletion = NO;
            [self.topImg.layer addAnimation:animation forKey:nil];
        }];
}

- (void)stopAnimating
{
    [UIView animateWithDuration:0.25 animations:^{
        self.alpha = 0;
    } completion:^(BOOL finished) {
        [self.topImg.layer removeAllAnimations];
    }];
}

@end

File: Swift/MSIM-Swift/Home/BFSparkLoadingView.swift
//
//  BFSparkLoadingView.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/7/29.
//

import UIKit
import MSIMSDK
import Kingfisher

class BFSparkLoadingView: UIView {

    public func beginAnimating() {
        
        alpha = 0
        UIView.animate(withDuration: 0.25) {
            self.alpha = 1
        } completion: { _ in
            let ani = CABasicAnimation(keyPath: "transform.rotation.z")
            ani.fromValue = NSNumber(floatLiteral: 0)
            ani.toValue = NSNumber(floatLiteral: Double.pi * 2)
            ani.duration = 3
            ani.autoreverses = false
            ani.fillMode = .forwards
            ani.repeatCount = Float.greatestFiniteMagnitude
            ani.isRemovedOnCompletion = false
            self.topImg.layer.add(ani, forKey: nil)
        }
    }
    
    public func stopAnimating() {
        
        UIView.animate(withDuration: 0.25) {
            self.alpha = 0
        } completion: { _ in
            self.topImg.layer.removeAllAnimations()
        }
    }
    
    private let bgImg = UIImageView(image: UIImage(named: "loadingbg"))
    
    private let topImg = UIImageView(image: UIImage(named: "loadingtop"))
    
    private let maskImg = UIImageView(image: UIImage(named: "loadingheadmask"))
    
    private var headImg = UIImageView()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        addSubview(bgImg)
        addSubview(topImg)
        addSubview(maskImg)
        
        headImg.contentMode = .scaleAspectFill;
        headImg.layer.cornerRadius = 4
        headImg.layer.masksToBounds = true
        if let uid = MSIMTools.sharedInstance().user_id,let me = MSProfileProvider().providerProfile(fromLocal: uid) {
            headImg.kf.setImage(with: URL(string: me.avatar))
        }
        addSubview(headImg)
        alpha = 0
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        bgImg.frame = CGRect(x: 0, y: 0, width: UIScreen.width, height: UIScreen.width)
        bgImg.center = center
        topImg.frame = CGRect(x: 0, y: 0, width: UIScreen.width, height: UIScreen.width)
        topImg.center = center
        maskImg.frame = CGRect(x: 0, y: 0, width: 97, height: 97)
        maskImg.center = center
        headImg.frame = CGRect(x: 0, y: 0, width: 41, height: 55)
        headImg.center = center
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}
